fs/fs.go

```diff
@@ -74,20 +74,79 @@ import (
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
 )
 
 var (
 	defaultIndexSelectionPolicy = SelectFirstPolicy
 	fusermountBin               = "fusermount"
+	preresolverQueueBufferSize  = 1024 // arbitrarily chosen buffer size
 )
 
+// Preresolver will resolve a number of layers in parallel,
+// up to the amount specified by MaxConcurrency.
+type preresolver struct {
+	queue chan func(context.Context) string
+	cache *sync.Map
+	smp   *semaphore.Weighted
```

Comment:
Is this a clear enough name for a semaphore? Not sure if there's generally a convention for what to call them in a struct

Reply:
`smp   *semaphore.Weighted` what does the weighted means here?

Reply:
It's an implementation of the semaphore in [golang's library](https://pkg.go.dev/golang.org/x/sync/semaphore#Weighted). It works effectively like a semaphore, but instead of only allowing one thread per `Acquire()`, it allows processes to take multiple threads instead.